pkg/metrics/metrics.go

```diff
@@ -119,8 +123,146 @@ func initOperatorService(port int32, portName string) (*v1.Service, error) {
 					Name: portName,
 				},
 			},
-			Selector: map[string]string{"name": operatorName},
+			Selector: label,
 		},
 	}
+
+	ownRef, err := getPodOwnerRef(ctx, client, namespace)
+	if err != nil {
+		return nil, err
+	}
+	service.SetOwnerReferences([]metav1.OwnerReference{*ownRef})
+
 	return service, nil
 }
+
+func getPodOwnerRef(ctx context.Context, client crclient.Client, ns string) (*metav1.OwnerReference, error) {
+	// Get current Pod the operator is running in
+	pod, err := k8sutil.GetPod(ctx, client, ns)
+	if err != nil {
+		return nil, err
+	}
+	podOwnerRefs := &metav1.OwnerReference{
+		APIVersion: "core/v1",
+		Kind:       "Pod",
+		Name:       pod.Name,
+		UID:        pod.UID,
+		Controller: &trueVar,
+	}
+
+	// Get Owner that the Pod belongs to
+	ownerRef := metav1.GetControllerOf(pod)
+	finalOwnerRef, err := findFinalOwnerRef(ctx, client, ns, ownerRef)
+	if err != nil {
+		return nil, err
+	}
+	if finalOwnerRef != nil {
+		return finalOwnerRef, nil
+	}
+
+	// Default to returning Pod as the Owner
+	return podOwnerRefs, nil
+}
+
+// findFinalOwnerRef tries to locate the final controller/owner based on the owner reference provided.
+func findFinalOwnerRef(ctx context.Context, client crclient.Client, ns string, ownerRef *metav1.OwnerReference) (*metav1.OwnerReference, error) {
```

Comment:
I really think that this function has to handle `DeploymentConfigs`

Reply:
I agree, but I think we should try to do it as generically as possible. I wonder if we could consolidate the switch statement cases for known types to be:

```go
case "ReplicaSet", "ReplicationController":
	rsrc := &unstructured.Unstructured{}
	key := crclient.ObjectKey{Namespace: ns, Name: ownerRef.Name}
	if err := client.Get(ctx, key, rsrc); err != nil {
		return nil, err
	}
	rsrcOwner := metav1.GetControllerOf(rsrc)

	// If we find an owner for the RS/RC, return that.
	// Otherwise, just return the ownerRef directly.
	if rsrcOwner != nil {
		return rsrcOwner, nil
	}
	return ownerRef, nil
case "DaemonSet", "StatefulSet", "Job":
	// I think we can just return this directly.
	// Any reason to fetch the object first?
	return ownerRef, nil
```

I think this would cover Deployments and DeploymentConfigs generically, and it would be a bit more future-proof since we're being less explicit about some of the types we care about.

Reply:
I agree with @joelanford. A generic approach will reduce code significantly.

Reply:
``` 	// I think we can just return this directly.
	// Any reason to fetch the object first?
	return ownerRef, nil
```

Agreed! 🤦‍♀️

Reply:
This would then return the RS/RC as the owner directly, above we said we want to avoid doing that? Did I misunderstand your suggestion above maybe https://github.com/operator-framework/operator-sdk/pull/1037#issuecomment-459498255
```go
// If we find an owner for the RS/RC, return that.
	// Otherwise, just return the ownerRef directly.
	if rsrcOwner != nil {
		return rsrcOwner, nil
	}
```
The reason was to avoid returning RS/RC as the owner and to avoid returning non k8s controllers owners as well.

Reply:
@LiliC I think if the `ReplicaSet` or `ReplicationController` has an owner we should return its owner, regardless of the Kind of the owner. That would cover the cases for `Deployment` and `DeploymentConfig` implicitly and it would mean other custom or future native K8s or Openshift types that own and manage `ReplicaSets` and `ReplicationControllers` would be supported without any code changes.

If the `ReplicaSet` or `ReplicationController` does not have an owner, I think it makes the most sense to use the `ReplicaSet` or `ReplicationController` as the owner directly, since it would be the highest level Kind we found from our traversal of the owner references from the pod.